README.md
# react-asyncbutton

![npm](https://img.shields.io/npm/v/react-asyncbutton.svg) ![license](https://img.shields.io/npm/l/react-asyncbutton.svg) ![github-issues](https://img.shields.io/github/issues/revolunet/react-asyncbutton.svg)

async button to call urls with visual feedback

![demo](./examples/asyncbutton.gif)

## Install

`npm install --save-dev react-asyncbutton`

## Usage

```html
<AsyncButton url={ url } onSuccess={ this.onSuccess }/>
```

See [examples](./examples)

## Scripts  

 - **npm run test** : `rackt test`
 - **npm run test-chrome** : `./node_modules/.bin/testem -l Chrome`
 - **npm run start** : `rackt server`

## Dependencies

Package | Version | Dev
--- |:---:|:---:
[babel-runtime](https://www.npmjs.com/package/babel-runtime) | 6.1.4 | ✖
[browser-request](https://www.npmjs.com/package/browser-request) | ^0.3.3 | ✖
[babel-runtime](https://www.npmjs.com/package/babel-runtime) | ^6.1.4 | ✔
[node-readme](https://www.npmjs.com/package/node-readme) | 0.1.9 | ✔
[rackt-cli-revolunet](https://www.npmjs.com/package/rackt-cli-revolunet) | ^0.5.10 | ✔
[react](https://www.npmjs.com/package/react) | ^0.14 | ✔


## Contributing

Contributions welcome; Please submit all pull requests the against master branch. If your pull request contains JavaScript patches or features, you should include relevant unit tests. Please check the [Contributing Guidelines](contributng.md) for more details. Thanks!

## Author

Julien Bouquillon <julien@revolunet.com>> undefined

## License

 - **MIT** : http://opensource.org/licenses/MIT


CHANGELOG.md
v1.1.0 - Tue, 24 May 2016 00:35:44 GMT
--------------------------------------

- 


v1.0.1 - Wed, 11 Nov 2015 15:13:01 GMT
--------------------------------------

- 


v1.0.0 - Wed, 11 Nov 2015 15:11:09 GMT
--------------------------------------

- 




lib/components/__tests__/ReactAsyncbutton-test.js
import React from 'react';
import expect from 'expect';
import { createRenderer } from 'react-addons-test-utils';
import expectJSX from 'expect-jsx';
expect.extend(expectJSX);

import ReactAsyncbutton from '../ReactAsyncbutton';

describe('ReactAsyncbutton', () => {
  it('works', () => {
    let renderer = createRenderer();
    renderer.render(<ReactAsyncbutton name='John' click={ () => {} } />);
    let actualElement = renderer.getRenderOutput();
    let expectedElement = <div onClick={ () => {} }>John</div>;
    expect(actualElement).toEqualJSX(expectedElement);
  });

  it('has a click property', () => {
    let renderer = createRenderer();
    let hasClicked = false;
    let click = () => hasClicked = true;
    renderer.render(<ReactAsyncbutton name='John' click={ click } />);
    renderer.getRenderOutput().props.onClick();
    expect(hasClicked).toBe(true);
  });
});


lib/components/ReactAsyncbutton.js
import React, { Component, PropTypes } from 'react';

import request from 'browser-request';

// button that fire a get request and display progress (aka poor man ladda button)
export default class AsyncButton extends Component {
  static propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string,
    labels: PropTypes.object
  };
  static defaultProps = {
    title: '',
    labels: {
      default: 'Click Me',
      pending: 'Updating',
      success: 'Done!',
      error: 'Error'
    }
  };
  static STATUS = {
    default: 'default',
    pending: 'pending',
    success: 'success',
    error: 'error'
  }
  constructor(...args) {
    super(...args);
    this.state = {
      status: ButtonAsync.STATUS.default
    };
    this.onClick = ::this.onClick;
  }
  onClick() {
    if (this.state.status === ButtonAsync.STATUS.pending) {
      return;
    }
    if (this.state.status === ButtonAsync.STATUS.success || this.state.status === ButtonAsync.STATUS.error) {
      this.setState({
        status: ButtonAsync.STATUS.default
      });
      return;
    }
    this.setState({
      status: ButtonAsync.STATUS.pending
    }, () => {
      // do req + success / failure
      request({
        uri: this.props.url,
        method: 'GET',
        json: true,
        timeout: 30000
      }, (err, res) => {
        if (err) {
          console.log('error', err);
          this.setState({
            status: ButtonAsync.STATUS.error
          });
        } else {
          console.log('success', res);
          this.setState({
            status: ButtonAsync.STATUS.success
          });
        }
      });
    });
  }
  render() {
    let buttonStyle = {
      fontSize: 14,
      borderRadius: 5,
      background: '#eee',
      minWidth: 130
    };
    let disabled = (this.state.status === ButtonAsync.STATUS.pending);
    let label = this.props.labels[this.state.status];
    return (<button
            title={ this.props.title }
            style={ buttonStyle }
            disabled={ disabled }
            onClick={ this.onClick }
           >{ label }
           </button>);
  }
}


lib/components/AsyncButton.js
import React, { Component, PropTypes } from 'react';

import request from 'browser-request';

// button that fire a get request and display progress (aka poor man ladda button)
export default class AsyncButton extends Component {
  static propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string,
    labels: PropTypes.object,
    style: PropTypes.object,
    className: PropTypes.string,
    onError: PropTypes.func,
    onSuccess: PropTypes.func
  };
  static defaultProps = {
    title: '',
    className: '',
    style: {},
    labels: {
      default: 'Click Me',
      pending: 'Updating',
      success: 'Done!',
      error: 'Error'
    }
  };
  static STATUS = {
    default: 'default',
    pending: 'pending',
    success: 'success',
    error: 'error'
  }
  constructor(...args) {
    super(...args);
    this.state = {
      status: AsyncButton.STATUS.default
    };
    this.onClick = ::this.onClick;
  }
  onClick() {
    if (this.state.status === AsyncButton.STATUS.pending) {
      return;
    }
    if (this.state.status === AsyncButton.STATUS.success || this.state.status === AsyncButton.STATUS.error) {
      this.setState({
        status: AsyncButton.STATUS.default
      });
      return;
    }
    this.setState({
      status: AsyncButton.STATUS.pending
    }, () => {
      // do req + success / failure
      request({
        uri: this.props.url,
        method: 'GET',
        json: true,
        timeout: 30000
      }, (err, res) => {
        if (err) {
          this.setState({
            status: AsyncButton.STATUS.error
          });
          if (this.props.onError) {
            this.props.onError(err);
          }
        } else {
          this.setState({
            status: AsyncButton.STATUS.success
          });
          if (this.props.onSuccess) {
            this.props.onSuccess(res);
          }
        }
      });
    });
  }
  render() {
    let style = {
      fontSize: 14,
      borderRadius: 5,
      background: '#eee',
      minWidth: 130,
      ...this.props.style
    };
    let className = this.props.className;
    let disabled = (this.state.status === AsyncButton.STATUS.pending);
    let label = this.props.labels[this.state.status];
    return (<button
            className={ className }
            title={ this.props.title }
            style={ style }
            disabled={ disabled }
            onClick={ this.onClick }
           >{ label }
           </button>);
  }
}
